### PR TITLE
[Feature] Support add/drop filed for struct column in share-data mode

### LINKED ISSUE #46452: Support Add/drop Field for struct column
- [x] https://github.com/StarRocks/starrocks/pull/49192
- [x] https://github.com/StarRocks/starrocks/pull/48430
- [x] https://github.com/StarRocks/starrocks/pull/47590
- [x] https://github.com/StarRocks/starrocks/pull/47581
- [x] https://github.com/StarRocks/starrocks/pull/47110
- [x] https://github.com/StarRocks/starrocks/pull/47497
- [x] https://github.com/StarRocks/starrocks/pull/47206
- [x] https://github.com/StarRocks/starrocks/pull/47217
- [x] https://github.com/StarRocks/starrocks/pull/46619
- [x] https://github.com/StarRocks/starrocks/pull/47109
- [x] https://github.com/StarRocks/starrocks/pull/46451

## Feature request
Support add/drop field for struct column

**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by sevev**
## Why I'm doing:
Support add/drop field for struct column in shared-data mode.

## What I'm doing:

Fixes https://github.com/StarRocks/starrocks/issues/46452

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [x] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: gensrc/proto/lake_types.proto [hunk 1]
```diff
@@ -141,6 +141,8 @@ message TabletMetadataPB {
     optional TabletSchemaPB source_schema = 14;
     optional PersistentIndexSstableMetaPB sstable_meta = 15;
     optional DeltaColumnGroupMetadataPB dcg_meta = 16;
+    map<int64, TabletSchemaPB> historical_schema = 17;
```
Comment (starrocks-xupeng) [inline gensrc/proto/lake_types.proto:144]: add coment for int64_t

File: gensrc/proto/lake_types.proto [hunk 2]
```diff
@@ -141,6 +141,8 @@ message TabletMetadataPB {
...
+    repeated int64 rowset_schema_id = 18;
```
Comment (starrocks-xupeng) [inline gensrc/proto/lake_types.proto:145]: add comment for vector rowset_schema_id

File: be/src/storage/lake/txn_log_applier.cpp [hunk 3]
```diff
@@ -197,6 +228,7 @@ class PrimaryKeyTxnLogApplier : public TxnLogApplier {
             !op_write.rowset().has_delete_predicate()) {
             return Status::OK();
         }
+        Status st;
```
Comment (wyb) [inline be/src/storage/lake/txn_log_applier.cpp:231]: not used

File: be/src/storage/lake/meta_file.cpp [hunk 4]
```diff
@@ -210,6 +213,25 @@ void MetaFileBuilder::apply_opcompaction(const TxnLogPB_OpCompaction& op_compact
         uint32_t id;
         bool operator()(const uint32_t rowid) const { return rowid == id; }
     };
+
+    struct RowsetFinder {
+        int64_t id;
+        bool operator()(const RowsetMetadata& r) const { return r.id() == id; }
+    };
+    // get input rowset index
+    std::set<uint32_t> input_rowset_index;
```
Comment (wyb) [inline be/src/storage/lake/meta_file.cpp:222]: better to use `unordered_set`

File: be/src/common/config.h [hunk 5]
```diff
@@ -1356,4 +1356,7 @@ CONF_Int32(python_worker_expire_time_sec, "300");
 CONF_mBool(enable_pk_strict_memcheck, "true");
 CONF_mBool(skip_lake_pk_preload, "false");
 
+// Experimental feature, this configuration will be removed after testing is complete.
+CONF_mBool(enable_alter_struct, "true");
```
Comment (starrocks-xupeng) [inline be/src/common/config.h:1360]: `lake_enable_alter_struct` ?

File: gensrc/proto/lake_types.proto (hunk 2 above)
Comment (sevev) [reply gensrc/proto/lake_types.proto:148]: > add comment for vector rowset_schema_id

done

File: gensrc/proto/lake_types.proto [hunk 6]
```diff
@@ -141,6 +141,11 @@ message TabletMetadataPB {
     optional TabletSchemaPB source_schema = 14;
     optional PersistentIndexSstableMetaPB sstable_meta = 15;
     optional DeltaColumnGroupMetadataPB dcg_meta = 16;
+    // Keep the tablet schema before each Alter operation
+    // If no rowset use the tablet schema any more, it will be gc by compaction
+    map<int64, TabletSchemaPB> historical_schema = 17;
+    // Keep tablet schema id for each rowset
+    repeated int64 rowset_schema_id = 18;
```
Comment (wyb) [inline gensrc/proto/lake_types.proto:148]: better to use map, `rowset id -> schema id`

File: gensrc/proto/lake_types.proto [hunk 7]
```diff
@@ -141,6 +141,11 @@ message TabletMetadataPB {
     optional TabletSchemaPB source_schema = 14;
     optional PersistentIndexSstableMetaPB sstable_meta = 15;
     optional DeltaColumnGroupMetadataPB dcg_meta = 16;
+    // Keep the tablet schema before each Alter operation
+    // If no rowset use the tablet schema any more, it will be gc by compaction
+    map<int64, TabletSchemaPB> historical_schema = 17;
```
Comment (wyb) [inline gensrc/proto/lake_types.proto:146]:
```suggestion
    map<int64, TabletSchemaPB> historical_schemas = 17;
```

File: gensrc/proto/lake_types.proto [hunk 8]
```diff
@@ -141,6 +141,11 @@ message TabletMetadataPB {
...
+    // rowset_id -> schema_id
+    map<uint32, int64> rowset_schema_id = 18;
```
Comment (wyb) [inline gensrc/proto/lake_types.proto:148]:
```suggestion
    map<uint32, int64> rowset_to_schema = 18;
```

File: be/src/storage/lake/meta_file.cpp [hunk 9]
```diff
@@ -135,6 +135,9 @@ void MetaFileBuilder::apply_opwrite(const TxnLogPB_OpWrite& op_write, const std:
         file_meta.set_name(orphan_file);
         _tablet_meta->mutable_orphan_files()->Add(std::move(file_meta));
     }
+    if (!_tablet_meta->rowset_schema_id().empty()) {
+        (*_tablet_meta->mutable_rowset_schema_id())[rowset->id()] = -1;
```
Comment (wyb) [inline be/src/storage/lake/meta_file.cpp:139]: why the value is -1, instead of current schema id?

File: be/src/storage/lake/meta_file.cpp [hunk 10]
```diff
@@ -271,6 +282,42 @@ void MetaFileBuilder::apply_opcompaction(const TxnLogPB_OpCompaction& op_compact
...
+            auto input_id = op_compaction.input_rowsets(i);
+            _tablet_meta->mutable_rowset_schema_id()->erase(input_id);
+        }
+
+        if (has_output_rowset) {
+            (*_tablet_meta->mutable_rowset_schema_id())[output_rowset_id] = output_rowset_schema_id;
+        }
+
+        std::set<int64_t> erase_id;
+        std::set<int64_t> schema_id;
```
Comment (wyb) [inline be/src/storage/lake/meta_file.cpp:307]:
```suggestion
        std::unordered_set<int64_t> schema_ids;
```

File: be/src/storage/lake/meta_file.cpp [hunk 11]
```diff
@@ -271,6 +282,42 @@ void MetaFileBuilder::apply_opcompaction(const TxnLogPB_OpCompaction& op_compact
...
+
+        for (auto& pair : _tablet_meta->historical_schema()) {
+            if (schema_id.find(pair.first) == schema_id.end()) {
+                erase_id.insert(pair.first);
+            }
+        }
+
+        for (auto id : erase_id) {
+            _tablet_meta->mutable_historical_schema()->erase(id);
+        }
```
Comment (wyb) [inline be/src/storage/lake/meta_file.cpp:320]: better to use `iterator` to delete and `schema_id set` can be removed

File: be/src/storage/lake/txn_log_applier.cpp [hunk 12]
```diff
@@ -44,9 +44,27 @@ Status apply_alter_meta_log(TabletMetadataPB* metadata, const TxnLogPB_OpAlterMe
             // But it will be remove from index cache after apply is finished
             (void)update_mgr->index_cache().try_remove_by_key(metadata->id());
         }
+        // update tablet meta
+        // 1. rowset_to_schema is empty, maybe upgrade from old version or first time to do fast ddl
+        // 2. rowset_to_schema is not empty, add the origin tablet schema into historical_schemas and update the
+        //    rowset_to_schema
```
Comment (wyb) [inline be/src/storage/lake/txn_log_applier.cpp:50]: update the comment

File: be/test/exprs/like_test.cpp [hunk 13]
```diff
@@ -38,7 +38,7 @@ class LikeTest : public ::testing::Test {
 public:
     TExprNode expr_node;
 };
-
+/*
```
Comment (wyb) [inline be/test/exprs/like_test.cpp:41]: recover?

**Review by wyb — DISMISSED** [review]

**Review by wyb — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Review by dirtysalt — APPROVED** [review]

**Review by HangyuanLiu — APPROVED** [review]

**Comment by sevev** [comment]
@mergify backport branch-3.3

**Pull request merged**